File: frontend/src/lib/destruction.ts
import { api } from './api'
import type { PagedResult } from './incomingMail'

export interface EligibilityDto { documentId: number; eligible: boolean; reasons: string[] }

export interface LegalHold {
  id: number; reason: string; scope: number
  documentId: number | null; folderId: number | null; orgUnitId: number | null
  placedByUserId: number; placedByName: string | null; placedAtUtc: string
  releasedByUserId: number | null; releasedAtUtc: string | null; isActive: boolean
}

export interface DestructionItem {
  id: number; documentId: number; documentNumber: string; documentTitle: string
  method: number; customMethod: string | null; checksumBefore: string | null; outcome: string | null
}

export interface DestructionRequest {
  id: number; status: string; reason: string; retentionBasisId: number | null
  requestedByUserId: number; requestedByName: string | null; requestedAtUtc: string
  approvedByUserId: number | null; approvedByName: string | null; approvedAtUtc: string | null
  executedByUserId: number | null; executedAtUtc: string | null
  decisionNote: string | null; scheduledForUtc: string | null; certificateId: number | null
  items: DestructionItem[]
}

export const METHOD_LABELS: Record<number, string> = {
  0: 'محو التشفير (Crypto-Shred)', 1: 'كتابة آمنة فوقية', 2: 'حذف + إبطال البصمة',
  3: 'تقطيع', 4: 'حرق', 5: 'تذويب', 6: 'إزالة مغناطيسية', 7: 'أخرى (حدّد الطريقة)',
}
export const METHOD_OTHER = 7

export const DESTRUCTION_STATUS_LABELS: Record<string, string> = {
  Draft: 'مسودة', PendingReview: 'قيد المراجعة', PendingApproval: 'بانتظار الاعتماد',
  Approved: 'معتمد', Rejected: 'مرفوض', Scheduled: 'مجدول', Executing: 'قيد التنفيذ',
  Completed: 'منفّذ', Cancelled: 'ملغى', Failed: 'فشل',
}

export const HOLD_SCOPE_LABELS: Record<number, string> = { 0: 'وثيقة', 1: 'مجلد', 2: 'وحدة تنظيمية', 3: 'استعلام' }

export const destruction = {
  eligibility: (documentId: number) =>
    api.get<EligibilityDto>(`/destruction/eligibility/${documentId}`).then((r) => r.data),
  list: (status?: number) =>
    api.get<PagedResult<DestructionRequest>>('/destruction/requests', { params: { status, pageSize: 50 } }).then((r) => r.data),
  get: (id: number) => api.get<DestructionRequest>(`/destruction/requests/${id}`).then((r) => r.data),
  create: (b: { documentIds: number[]; reason: string; method: number; customMethod?: string | null; scheduledForUtc?: string | null }) =>
    api.post<DestructionRequest>('/destruction/requests', b).then((r) => r.data),
  submit: (id: number) => api.post<DestructionRequest>(`/destruction/requests/${id}/submit`, {}).then((r) => r.data),
  approve: (id: number, note?: string | null) => api.post<DestructionRequest>(`/destruction/requests/${id}/approve`, { note }).then((r) => r.data),
  reject: (id: number, note?: string | null) => api.post<DestructionRequest>(`/destruction/requests/${id}/reject`, { note }).then((r) => r.data),
  cancel: (id: number) => api.post<DestructionRequest>(`/destruction/requests/${id}/cancel`, {}).then((r) => r.data),
  execute: (id: number, b: { stepUpPassword: string; physicalOfficer?: string | null; physicalWitness?: string | null }) =>
    api.post<DestructionRequest>(`/destruction/requests/${id}/execute`, b).then((r) => r.data),
  downloadCertificate: async (id: number) => {
    const res = await api.get(`/destruction/requests/${id}/certificate`, { responseType: 'blob' })
    const url = URL.createObjectURL(res.data as Blob)
    const a = document.createElement('a')
    a.href = url; a.download = `certificate-${id}.pdf`
    document.body.appendChild(a); a.click(); a.remove(); URL.revokeObjectURL(url)
  },
}

export interface DestructionMethodOption { id: number; label: string; sortOrder: number; isActive: boolean }

export const destructionMethods = {
  list: () => api.get<DestructionMethodOption[]>('/destruction/methods').then((r) => r.data),
  create: (b: { label: string; isActive: boolean }) => api.post<DestructionMethodOption>('/destruction/methods', b).then((r) => r.data),
  update: (id: number, b: { label: string; isActive: boolean }) => api.put<DestructionMethodOption>(`/destruction/methods/${id}`, b).then((r) => r.data),
  remove: (id: number) => api.delete(`/destruction/methods/${id}`),
}

export const legalHolds = {
  list: (activeOnly = false) => api.get<LegalHold[]>('/legal-holds', { params: { activeOnly } }).then((r) => r.data),
  place: (b: { reason: string; scope: number; documentId?: number | null; folderId?: number | null; orgUnitId?: number | null }) =>
    api.post<LegalHold>('/legal-holds', b).then((r) => r.data),
  release: (id: number) => api.post(`/legal-holds/${id}/release`, {}),
}

File: backend/src/Archiving.Domain/Entities/Destruction.cs
using Archiving.Domain.Common;

namespace Archiving.Domain.Entities;

public enum LegalHoldScope { Document = 0, Folder = 1, OrgUnit = 2, Query = 3 }

public enum DestructionStatus
{
    Draft = 0, PendingReview = 1, PendingApproval = 2, Approved = 3, Rejected = 4,
    Scheduled = 5, Executing = 6, Completed = 7, Cancelled = 8, Failed = 9
}

public enum DestructionMethod
{
    // Digital
    CryptoShred = 0, SecureOverwrite = 1, DeleteFixityVoid = 2,
    // Physical
    Shredding = 3, Incineration = 4, Pulping = 5, Degaussing = 6,
    // Free-text method described by the requester
    Other = 7
}

/// <summary>A legal/litigation hold that makes the in-scope records ineligible for destruction
/// until released. While any active hold matches a document, it can never be destroyed.</summary>
public class LegalHold : BaseEntity
{
    public string Reason { get; set; } = string.Empty;
    public LegalHoldScope Scope { get; set; }
    public long? DocumentId { get; set; }       // Scope = Document
    public long? FolderId { get; set; }         // Scope = Folder
    public long? OrgUnitId { get; set; }        // Scope = OrgUnit
    public string? QueryExpression { get; set; } // Scope = Query (reserved for later)
    public long PlacedByUserId { get; set; }
    public DateTime PlacedAtUtc { get; set; } = DateTime.UtcNow;
    public long? ReleasedByUserId { get; set; }
    public DateTime? ReleasedAtUtc { get; set; } // null => still active
}

/// <summary>A controlled, auditable request to destroy one or more records that have met retention.</summary>
public class DestructionRequest : BaseEntity
{
    public DestructionStatus Status { get; set; } = DestructionStatus.Draft;
    public string Reason { get; set; } = string.Empty;
    public long? RetentionBasisId { get; set; }  // RetentionPolicy authorizing disposal
    public long RequestedByUserId { get; set; }
    public DateTime RequestedAtUtc { get; set; } = DateTime.UtcNow;
    public long? ReviewedByUserId { get; set; }
    public DateTime? ReviewedAtUtc { get; set; }
    public long? ApprovedByUserId { get; set; }
    public DateTime? ApprovedAtUtc { get; set; }
    public long? ExecutedByUserId { get; set; }
    public DateTime? ExecutedAtUtc { get; set; }
    public string? DecisionNote { get; set; }
    public DateTime? ScheduledForUtc { get; set; }
    public long? CertificateId { get; set; }
    public long? WorkflowInstanceId { get; set; }
    public ICollection<DestructionItem> Items { get; set; } = new List<DestructionItem>();
}

/// <summary>A single document line on a destruction request, with its method and (post-execution) outcome.</summary>
public class DestructionItem : BaseEntity
{
    public long DestructionRequestId { get; set; }
    public long DocumentId { get; set; }
    public DestructionMethod Method { get; set; } = DestructionMethod.CryptoShred;
    public string? CustomMethod { get; set; }     // free-text method when Method == Other
    public string? ChecksumBefore { get; set; }  // captured at execution as proof-of-prior-existence
    public string? Outcome { get; set; }          // per-item result after execution
    public DestructionRequest Request { get; set; } = null!;
}

/// <summary>An admin-managed destruction method label, shown in the request form. The actual digital
/// technique is always crypto-shred/secure-overwrite; this catalog lets admins name the methods
/// (e.g. physical methods) used in their organization.</summary>
public class DestructionMethodOption : BaseEntity
{
    public string Label { get; set; } = string.Empty;
    public int SortOrder { get; set; }
    public bool IsActive { get; set; } = true;
}

/// <summary>The Certificate of Destruction (rendered to PDF/A in a later phase).</summary>
public class DestructionCertificate : BaseEntity
{
    public long DestructionRequestId { get; set; }
    public string CertificateNumber { get; set; } = string.Empty;
    public string? PdfStorageKey { get; set; }
    public DateTime IssuedAtUtc { get; set; } = DateTime.UtcNow;
}
